### PR TITLE
Port FV3-GSDCHEM-WW3 for GEFSv12 to WCOSS2

### PR DESCRIPTION
**Pull request by XianwuXue-NOAA**

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -1,12 +1,12 @@
...
 	path = FV3
-	url = gerrit:FV3
+	url = https://github.com/NOAA-EMC/fv3atm
 [submodule "FMS"]
 	path = FMS
 	url = https://github.com/NOAA-EMC/FMS
 [submodule "GSDCHEM"]
 	path = GSDCHEM
-	url = gerrit:GSDCHEM
+	url = https://github.com/NOAA-EMC/GSDCHEM.git
```
Comment (WalterKolczynski-NOAA) [inline .gitmodules]: Why does GSDCHEM have a .git extension and none of the others do?

**Comment by XianwuXue-NOAA** [comment]
@WalterKolczynski-NOAA @KateFriedman-NOAA @JessicaMeixner-NOAA Could you review my changes? One question: should I also modify modulefiles for hera and wcoss to make it build on these two machines? Because I changed "module load $(CONFDIR)/modules.nems" to ". $(CONFDIR)/modules.nems" to make XX_ver use default value or passing value from build.ver.

File: conf/configure.fv3.wcoss2 [hunk 2]
```diff
@@ -0,0 +1,164 @@
+ifdef InNemsMakefile
+include $(TOP)/conf/configure.nems.NUOPC
```
Comment (WalterKolczynski-NOAA) [inline conf/configure.fv3.wcoss2:2]: Why does this file not have any indentation?

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

**Comment by WalterKolczynski-NOAA** [comment]
As long as the Hera module still works, I'll leave whether or not to update it to you. This repo is a dead-end anyway.

I removed Kate from the reviewer list, as she isn't appropriate for this.

File: .gitmodules (hunk 1 above)
Comment (XianwuXue-NOAA) [reply .gitmodules]: Thanks. I will modify it later

**Comment by WalterKolczynski-NOAA** [comment]
> I changed "module load $(CONFDIR)/modules.nems" to ". $(CONFDIR)/modules.nems" to make XX_ver use default value or passing value from build.ver.

If you change it this way, you are no longer loading a module, you are sourcing a shell script. You'd hardly be the first, but these are supposed to be modulefiles.

**Comment by XianwuXue-NOAA** [comment]
Changing that way is to using default value of XX_ver or accept values from outside (like accept value from build.ver based on WCOSS2 new standard: "The build.ver file must be sourced before compiling the code"

**Comment by WalterKolczynski-NOAA** [comment]
The version file has to be sourced to provide the versions, but these are still modulefiles. There should be ways to access environment variables in the modulefile (see https://modules.readthedocs.io/en/latest/modulefile.html#mfcmd-getenv, I think)

File: conf/configure.fv3.wcoss2 (hunk 2 above)
Comment (XianwuXue-NOAA) [reply conf/configure.fv3.wcoss2:2]: I copied it from configure.fv3.wcoss_dell_p3 and  configure.fv3.wcoss_cray, to be consistent with the others, I didn't change the format. However, I am OK to reformat it.

**Comment by XianwuXue-NOAA** [comment]
> The version file has to be sourced to provide the versions, but these are still modulefiles. There should be ways to access environment variables in the modulefile (see https://modules.readthedocs.io/en/latest/modulefile.html#mfcmd-getenv, I think)

This way is also using on GFSv16  in fv3 and UPP. If we want to change it, I think we should take with Kate and Wen about this. And I think it is an easier way to accept values. You knew that I built it on dell and hera with minor changes for the corresponding modulefiles, it works for both machines. I can submit the changes to this PR.

**Comment by WalterKolczynski-NOAA** [comment]
> > The version file has to be sourced to provide the versions, but these are still modulefiles. There should be ways to access environment variables in the modulefile (see https://modules.readthedocs.io/en/latest/modulefile.html#mfcmd-getenv, I think)
> 
> This way is also using on GFSv16 in fv3 and UPP. If we want to change it, I think we should take with Kate and Wen about this. And I think it is an easier way to accept values. You knew that I built it on dell and hera with minor changes for the corresponding modulefiles, it works for both machines. I can submit the changes to this PR.

Rahul is already pushing to have those fixed too.

**Comment by XianwuXue-NOAA** [comment]
> > > The version file has to be sourced to provide the versions, but these are still modulefiles. There should be ways to access environment variables in the modulefile (see https://modules.readthedocs.io/en/latest/modulefile.html#mfcmd-getenv, I think)
> > 
> > 
> > This way is also using on GFSv16 in fv3 and UPP. If we want to change it, I think we should take with Kate and Wen about this. And I think it is an easier way to accept values. You knew that I built it on dell and hera with minor changes for the corresponding modulefiles, it works for both machines. I can submit the changes to this PR.
> 
> Rahul is already pushing to have those fixed too.

Could you send me the link? I'd like to see Rahul's fixes.

**Comment by WalterKolczynski-NOAA** [comment]
> > > > The version file has to be sourced to provide the versions, but these are still modulefiles. There should be ways to access environment variables in the modulefile (see https://modules.readthedocs.io/en/latest/modulefile.html#mfcmd-getenv, I think)
> > > 
> > > 
> > > This way is also using on GFSv16 in fv3 and UPP. If we want to change it, I think we should take with Kate and Wen about this. And I think it is an easier way to accept values. You knew that I built it on dell and hera with minor changes for the corresponding modulefiles, it works for both machines. I can submit the changes to this PR.
> > 
> > 
> > Rahul is already pushing to have those fixed too.
> 
> Could you send me the link? I'd like to see Rahul's fixes.

I don't think he's made any himself (yet), he's just trying to get them done.

**Comment by XianwuXue-NOAA** [comment]
> > > > > The version file has to be sourced to provide the versions, but these are still modulefiles. There should be ways to access environment variables in the modulefile (see https://modules.readthedocs.io/en/latest/modulefile.html#mfcmd-getenv, I think)
> > > > 
> > > > 
> > > > This way is also using on GFSv16 in fv3 and UPP. If we want to change it, I think we should take with Kate and Wen about this. And I think it is an easier way to accept values. You knew that I built it on dell and hera with minor changes for the corresponding modulefiles, it works for both machines. I can submit the changes to this PR.
> > > 
> > > 
> > > Rahul is already pushing to have those fixed too.
> > 
> > 
> > Could you send me the link? I'd like to see Rahul's fixes.
> 
> I don't think he's made any himself (yet), he's just trying to get them done.

OK. Thanks for letting me know.  I am frozen the code for the canned data test. If he finds another way within one or two weeks, I can follow his way, otherwise, I think it is OK to keep this way. I have tested it works. Kate and Wen also confirmed.

**Review by WalterKolczynski-NOAA — APPROVED** [review]
Approved with the caveat we might turn modulefiles back into modulefiles instead of disguised shell scripts.

**Comment by WalterKolczynski-NOAA** [comment]
Here's one Rahul did do. It is in lua instead of tcl: https://github.com/NOAA-EMC/obsproc/blob/develop/modulefiles/obsproc_wcoss2.lua

**Comment by XianwuXue-NOAA** [comment]
> Here's one Rahul did do. It is in lua instead of tcl: https://github.com/NOAA-EMC/obsproc/blob/develop/modulefiles/obsproc_wcoss2.lua

Great. I checked the link and found it use "obsproc_common.lua" to overwrite the modules. I will see whether Kate and Wen will use this way in GFSv16 or not. If they use this way, I will try this way to unify with the global-workflow.

**Review by JessicaMeixner-NOAA — APPROVED** [review]
WW3 code looks good

**Comment by XianwuXue-NOAA** [comment]
@WalterKolczynski-NOAA @JessicaMeixner-NOAA Thanks a lot for your review. If everything is OK, could you please merge this PR since I am not authorized to merge?

**Comment by XianwuXue-NOAA** [comment]
@WalterKolczynski-NOAA Could you please create a tag "gefs_v12.2.0"?

**Comment by WalterKolczynski-NOAA** [comment]
> @WalterKolczynski-NOAA Could you please create a tag "gefs_v12.2.0"?

Are you sure this is ready? I'd prefer to wait until you run everything from this hash and make sure it is finalized.

**Comment by XianwuXue-NOAA** [comment]
OK. I will do a test next week.

**Comment by XianwuXue-NOAA** [comment]
@WalterKolczynski-NOAA I have tested and results are identical. So could you create a tag "gefs_v12.2.0"? Thanks.

**Pull request merged**